src/libstd/sys/unix/fs_linux.rs

```diff
@@ -0,0 +1,680 @@
+thread_local! {
+    static HAS_COPY_FILE_RANGE: RefCell<bool> = RefCell::new(true);
+}
+
+fn copy_bytes(reader: &File, writer: &File, uspace: bool, nbytes: u64) -> io::Result<u64> {
+    HAS_COPY_FILE_RANGE.with(|cfr| {
+        loop {
+            if uspace || !*cfr.borrow() {
+                return copy_bytes_uspace(reader, writer, nbytes as usize)
+
+            } else {
+                let result = copy_bytes_kernel(reader, writer, nbytes as usize);
+
+                if let Err(ref err) = result {
+                    match err.raw_os_error() {
+                        Some(libc::ENOSYS) | Some(libc::EPERM) => {
+                            // Flag as unavailable and retry.
+                            *cfr.borrow_mut() = false;
+                            continue;
+                        }
+                        _ => {}
+
+                    }
+                }
+                return result;
+            }
+        }
+    })
+}
+
+
+/// Copy len bytes from whereever the descriptor cursors are set.
+fn copy_range(infd: &File, outfd: &File, uspace: bool, len: u64) -> io::Result<u64> {
+    let mut written = 0;
+    while written < len {
+        let result = copy_bytes(&infd, &outfd, uspace, len - written)?;
+        written += result;
+    }
+    Ok(written)
+}
+
+fn next_sparse_segments(fd: &File, pos: u64, len: u64) -> io::Result<(u64, u64)> {
+    let next_data = match lseek(fd, pos as i64, Wence::Data)? {
+        SeekOff::Offset(off) => off,
+        SeekOff::EOF => len
+    };
+    let next_hole = match lseek(fd, next_data as i64, Wence::Hole)? {
+        SeekOff::Offset(off) => off,
+        SeekOff::EOF => len
+    };
+
+    Ok((next_data, next_hole))
+}
+
+fn copy_sparse(infd: &File, outfd: &File, uspace: bool, len: u64) -> io::Result<u64> {
+    allocate_file(&outfd, len)?;
+
+    let mut pos = 0;
+
+    while pos < len {
+        let (next_data, next_hole) = next_sparse_segments(infd, pos, len)?;
+        lseek(infd, next_data as i64, Wence::Set)?;
+        lseek(outfd, next_data as i64, Wence::Set)?;
+
+        let _written = copy_range(infd, outfd, uspace, next_hole - next_data)?;
+        pos = next_hole;
+    }
+
+    Ok(len)
+}
+
+
+fn copy_parms(in_meta: &Metadata, out_meta: &Metadata) -> io::Result<(bool, bool)> {
+    let is_sparse = in_meta.st_blocks() < in_meta.st_size() / in_meta.st_blksize();
+    let is_xmount = in_meta.st_dev() != out_meta.st_dev();
+    Ok((is_sparse, is_xmount))
+}
+
+
+pub fn copy(from: &Path, to: &Path) -> io::Result<u64> {
+    if !from.is_file() {
+        return Err(Error::new(ErrorKind::InvalidInput,
+                              "the source path is not an existing regular file"))
+    }
+
+    let infd = File::open(from)?;
+    let outfd = File::create(to)?;
+    let in_meta = infd.metadata()?;
+    let out_meta = outfd.metadata()?;
+
+    let (is_sparse, is_xmount) = copy_parms(&in_meta, &out_meta)?;
+    let uspace = is_xmount;
```

Comment:
This check may soon be too strict in new kernels, [cross-device clone_file_range might become supported](https://patchwork.kernel.org/patch/10656653/) e.g. between NFS mounts or multiple subvolumes of btrfs systems.

Instead one should always try `copy_file_range` and fall back to regular copy on the first `EXDEV`.